File: normalize.js
inlets = 1;
outlets = 1;

var input = [];


function list()
{
	switch (inlet) {
		case 0:
			input = arrayfromargs(arguments);
			break;
	}
	bang();
}

function bang()
{
    var sum = 0;
    var min = Infinity;
    var max = -Infinity;
    for (var i = 0; i < input.length; i++) {
        sum += input[i];
        if ((input[i]) < min) {
            min = (input[i]);
        }
        if ((input[i]) > max) {
            max = (input[i]);
        }
    }
    var avg = sum/input.length;

    var deviation = Math.max(max-avg, avg-min);

    var output = [];

    for (var i = 0; i < input.length; i++) {
        output.push((((input[i]-avg)/deviation)+1)/2);
    }
    outlet(0, output);
}

File: normalize.java
import com.cycling74.max.*;

public class normalize extends MaxObject {

    float[] input;

    normalize() {
        declareInlets(new int[] { DataTypes.LIST });
        declareOutlets(new int[] { DataTypes.LIST });
    }

    protected void list(final Atom[] a) {
        input = Atom.toFloat(a);
        bang();
    }

    protected void bang() {
        float sum = 0;
        float min = Float.POSITIVE_INFINITY;
        float max = Float.NEGATIVE_INFINITY;
        for (float f : input) {
            sum += f;
            if (f < min) {
                min = f;
            } else if (f > max) {
                max = f;
            }
        }
        float avg = sum / input.length;
        float deviation = Math.max(max - avg, avg - min);
        float[] output = new float[input.length];

        for (int i = 0; i < output.length; i++) {
            output[i] = (((input[i] - avg) / deviation) + 1.f) / 2.f;
        }

        outlet(0, output);
    }
}
